core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java

```diff
@@ -431,6 +431,7 @@ protected RelOptPlanner createPlanner(
     if (CalciteSystemProperty.ENABLE_COLLATION_TRAIT.value()) {
       planner.addRelTraitDef(RelCollationTraitDef.INSTANCE);
     }
+    planner.setTopDownOpt(prepareContext.config().topDownOpt());
```

Comment:
Hmm... This will always override of config value from CalciteSystemProperty when CalcitePrepareImpl.java is used. We need to think about it.

Reply:
Sorry, I mean here will overwrite. :)

Reply:
My current understanding is, we want to make sure Top-down optimization does not affect existing implementation. 

Then, not sure if you are aware of that, the value in `saffron.properties` always overwrite `CalciteSystemProperty.TOPDOWN_OPT`, thus `CalciteSystemProperty.TOPDOWN_OPT` is always true.

Thus, top-down optimizations are enabled by default in many tests. And the only way to update it is calling `planner.setTopDownOpt`.

This line is trying to disable top-down opt by default for JDBC code path (e.g. JDBC tests). Without this linke, top-down opt is enabled and ~40 tests are failing due to trait propagation implementation in this PR.

So do I understand correctly that top-down opt should be implemented right now in a relative isolated environment? Can we set the value in `saffron.properties` as false to turn off top-down opt by default? Is there better suggestions?

Reply:
Sorry for not being clear. `CalciteSystemProperty. TOPDOWN_OPT` reads data from `saffron.properties`, so it is intended to be globally true, because I wanted to make sure it passed **all** the tests. But now since we are going to add trait propagation for all the operators, and in next release, we will add top-down rule apply, there will be more plan diffs, so I think now we can turn if off in `saffron.properties` in this PR.

When top-down rule apply is committed into master, we need to come up with a solution to test the whole test suites with top-down on by default, but allow different plans.

Reply:
Sounds good. 

I set top-down opt as false in `saffron.properties`. Then made `CalciteConnectionProperty#TOPDOWN_OPT`'s default value is set by `CalciteSystemProperty.TOPDOWN_OPT.value()`, and I still keep this line code.

So it will achieve:
1. saffron.properties still can control top-down opt globally. 
2. Leave a room for people who wants to turn off top-down opt through Calcite connections.